src/ranch_conns_sup.erl

```diff
@@ -175,6 +181,11 @@ loop(State=#state{parent=Parent, ref=Ref, conn_type=ConnType,
 		{set_protocol_options, Opts2} ->
 			loop(State#state{opts=Opts2},
 				CurConns, NbChildren, Sleepers);
+		{activate_alarm, AlarmRef} ->
+			State2 = trigger_alarm(State#state{alarm_active=true}, CurConns),
+			loop(State2, CurConns, NbChildren, Sleepers);
+		{activate_alarm, _} ->
+			loop(State, CurConns, NbChildren, Sleepers);
```

Comment:
This catch all clause doesn't seem to be called currently because `alarm_ref` only gets updated on init and when `set_transport_options` is called.

Reply:
Remembered what this was about on the way home :) Maybe I should start keeping a diary, "What was I thinking when..." ;)

The name `alarm_ref` is probably misleading, as it's not (only) about the alarm feature and not necessarily a reference, though it's the most obvous choice. `transopts_tag` would probably be a better name. Decide after reading the rest.

So, the alarm settings should be changeable at runtime, like the other things you can change with `set_transport_options`. Change the treshold, change the cooldown period, turn off alarms altogether, etc. The most simple and trouble-free way is to unconditionally _enable_  the alarm (in other words, cancel the potential cooldown) when transport options get changed. This spares us a lot of gymnastics, and is a solid solution that I think anyone can live with.

But there is this problem: if the alarm is currently in cooldown, there is a timer running to send a re-activation message. If we enable the alarm "manually" as part of the transport options change, if the connection count is still over treshold, the alarm will get triggered right away, and in the end start another timer for cooling down from there, oblivious of the already running one. The two will then prematurely end the cooldown phases initiated by the respective other. If you change transport options again, you may end up with three, four, five... timer. It all settles down by itself once the connection count falls below the treshold for a while, but until then...

We could put the timer ref in the state in order to cancel it when we change transport options, but it may so happen that the timer expires in just that moment, the message is sent, and we can't take it back. I also have an inkling that there could be further problems if the alarm gets deactivated (set to "no alarm") and back in a short time.

The (solid and simple, I think) solution I came up with is to tag the transport options and include that tag in the reactivation message. If they match (the upper clause), the transport options have not changed since the timer was created, so the message is valid and the cooldown ends. If they _do_ change, however, a new tag is created, and the tag in the message doesn't match the tag in the state (the lower catchall clause), so the reactivation message is invalid (the timer was started by using old transport options) and gets ignored.

Actually, it would have worked (and probably be more obvious) if I had tagged each timed message anew and carried that tag in the state. Tagging the transport options instead amounts to much the same thing, it just saves the `make_ref` on timer creation, and may potentially be useful for other things in the future.

Reply:
If there's an existing cooldown timer and options change and you need to start a new timer, then you should probably not start a new timer? It's not like the alarm state changed or anything.

If there's an existing cooldown timer and options change and you need to stop the alarm, then you should cancel the existing timer and flush the message out, if any. Done in that order there is no race condition I think.

Rather than using a custom tag I would recommend using `start_timer` as well, it's the same except the `TimerRef` will be in the message and can be matched against. And that `TimerRef` should be kept in the state for the aforementioned cancel.

Reply:
> If there's an existing cooldown timer and options change and you need to start a new timer, then you should probably not start a new timer? It's not like the alarm state changed or anything.

I was thinking mostly about a change to the cooldown period here. Using extremes for illustration purposes, assume you have the cooldown set to 1 hour and want to change it to 1 second, then just going with the existing timer you may have to wait for that hour before the change really takes effect, in the worst case.

We could cancel the running timer, see how much time it had left, deduce from the old alarm transport options how long it had been running, see how that relates to the new cooldown period, and schedule the next (first after change) accordingly. That seemed comparatively complicated, and it will only work as intended once per change between cooldowns. So my idea here was to just start "fresh".

Also, there is a gray area where I'm not sure what _I_ would expect, much less what others would. If you are in cooldown and change the treshold to above the current connection count, would I want to be alerted as soon as the new treshold is reached, or would I want to wait for the leftover cooldown to expire? Again, starting fresh seemed to be the best option.

> If there's an existing cooldown timer and options change and you need to stop the alarm, then you should cancel the existing timer and flush the message out, if any.

Oh, right, you _can_ do that :) Didn't think of that.

> Done in that order there is no race condition I think.
> 
> Rather than using a custom tag I would recommend using `start_timer` as well, it's the same except the `TimerRef` will be in the message and can be matched against. And that `TimerRef` should be kept in the state for the aforementioned cancel.

Ok, good idea I think. Will see to that tomorrow then, as well as whatever you decide in relation to the above paragraph.

Reply:
Something based on http://erlang.org/doc/man/erlang.html#read_timer-2 is probably the best since you can easily know how much time passed, (re)trigger the alarm if necesssary, recreate a timer if necessary. It'll work regardless of how many times you change options.

I think about the gray area that the only concern is whether an alarm is already in effect. If you're in cooldown, you're in cooldown, unless (as noted above) the cooldown time changes.

Reply:
Ok, it's decided then :)

Reply:
What happened to this discussion?

Reply:
This specific one? It's settled I guess ^^;

If you mean the PR in general, I'm currently a bit busy, but started reworking the upgrade suite so I can make this fit in.

Reply:
Yes this specific discussion. I don't see `read_timer` in the code so I'm not sure it was done or if it simply became unnecessary?

Reply:
It is unnecessary. The alarm timers must only be accessed when the alarms are changed (via `set_transport_options`), and then only if an alarm is removed or it's cooldown changed:

* if a new alarm is added, it starts in active (non-cooldown) state anyway, so no need to fiddle with a timer
* if an alarm is removed, the associated cooldown timer, (if it is in cooldown at all), is cancelled. If that timer did not exist any more (`erlang:cancel_timer` returned `false`), it is assumed that it has just triggered, and the corresponding `timeout` message is flushed out
* if the cooldown period of an existing alarm is changed, same as above, but a new timer will be started taking into account the cooldown that has already expired on the old timer, making the transition "seamless"

This all happens in the `set_alarm_option/3`, `merge_alarms/2,3`, `adapt_alarm_timer/4` and `cancel_alarm_reactivation_timer/1` functions further down in the code.